contracts/ERC721Collateralizer.sol

```diff
@@ -0,0 +1,426 @@
+        address indexed debtor,
+        address erc721Contract,
+        uint256 tokenID
+    );
+
+    event CollateralSeized(
+        bytes32 indexed agreementID,
+        address indexed beneficiary,
+        address erc721Contract,
+        uint256 tokenID
+    );
+
+    modifier onlyAuthorizedToCollateralize() {
+        require(collateralizationPermissions.isAuthorized(msg.sender));
+        _;
+    }
+
+    function ERC721Collateralizer(
+        address _contractRegistry,
+        address _tokenRegistry,
+        address _cryptoKittiesContract
+    ) public {
+        ContractRegistry contractRegistry = ContractRegistry(_contractRegistry);
+
+        // Get the DebtRegistry contract from the ContractRegistry.
+        debtRegistry = contractRegistry.debtRegistry();
+        // Instantiate a special token registry for ERC721s.
+        tokenRegistry = ERC721TokenRegistry(_tokenRegistry);
+        // Keep track of the CryptoKitties contract, for specific workarounds.
+        cryptoKittiesContract = _cryptoKittiesContract;
+    }
+
+    /**
+     * Transfers collateral from the debtor to the current contract, as custodian.
+     *
+     * @param agreementId bytes32 The debt agreement's ID
+     * @param debtor address The owner of the asset being collateralized
+     */
+    function collateralize(
+        bytes32 agreementId,
+        address debtor
+    )
+        public
+        onlyAuthorizedToCollateralize
+        whenNotPaused
+        returns (bool _success)
+    {
+        // The address of the ERC721 contract.
+        address collateralTokenAddress;
+        // The token id, associating the collateral token and the given ERC721 contract.
+        uint256 collateralTokenID;
+        // The terms contract according to which this asset is being collateralized.
+        TermsContract termsContract;
+
+        // Fetch all relevant collateralization parameters
+        (
+            collateralTokenAddress,
+            collateralTokenID,
+            termsContract
+        ) = retrieveCollateralParameters(agreementId);
+
+        require(termsContract == msg.sender);
+        require(collateralTokenAddress != address(0));
+
+        /*
+            Ensure that the agreement has not already been collateralized.
+
+            If the agreement has already been collateralized, this check will fail
+            because any valid collateralization must have some sort of valid
+            address associated with it as a debtor.  Given that it is impossible
+            to send transactions from address 0x0, this check will only fail
+            when the agreement is already collateralized.
+        */
+        require(agreementToDebtor[agreementId] == address(0));
+
+        ERC721 erc721token = ERC721(collateralTokenAddress);
+        address custodian = address(this);
+
+        // Ensure that the debtor is the owner of this token ID.
+        require(erc721token.ownerOf(collateralTokenID) == debtor);
```

Comment:
Again, if we're being gas-conscious, this is a redundant check (721 should fail on `transferFrom` if the debtor does not own the token)